File: letter_case_permutation.py
"""
Given a string find all case permutations of a given letter

Time Complexity : O(2^N)
Space Complexity : O(N)

"""
class Solution(object):
    def is_number(self, s):
        try:
            int(s)
            return True
        except ValueError:
            return False
    
    def letterCasePermutation(self, s):
        if s == "":
            return [""]
        r = self.letterCasePermutation(s[1:])
        if self.is_number(s[0]):
            return [s[0] + i for i in r]
        else:
            return [s[0].upper() + i for i in r] + [s[0].lower() + i for i in r]

File: LetterCasePermutation.java
// Time Complexity : O(n)
// Space Complexity : O(n) Recursion Stack Space
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

import java.util.ArrayList;
import java.util.List;

public class LetterCasePermutation {
	class Solution {
	    public List<String> letterCasePermutation(String S) {
	        List<String> list = new ArrayList<>();
	        if(S == null || S.length() == 0){
	            list.add("");
	            return list;
	        }
	        List<String> result = letterCasePermutation(S.substring(1));
	        for(String s: result){
	            char c = S.charAt(0);
	            if(Character.isLetter(c)){
	                list.add(Character.toUpperCase(c) + s);
	                list.add(Character.toLowerCase(c) + s);
	            } else {
	                list.add(c + s);
	            }
	        }
	        return list;
	    }
	}
}
